### PR TITLE
Scan result to single numerical file

### PR DESCRIPTION
**Pull request by mlincett**
Currently, all scripts that use the scan result (plotting, finding best LLH, etc.) parse the cache directory of an event (few hundreds of MiBs in size) to obtain the required numerical values (notably the `llh` is used to determine the fit quality for the position of the pixel).

This PR introduces a `ScanResult` class that can be instantiated by reading an event from a cache directory (see associated `save_scan_result.py`) and saved to a `numpy` file (`.npz`) holding a dictionary of structured arrays (one per `nside`).

The same class can be instantiated by loading the `numpy` file itself and two `ScanResult` instances can be compared with either `__eq__` / `==` or `is_close()` methods.

In other words, **we make it easy and efficient to save and compare scan results** without having to load a huge cache directory. In the future, the class can be augmented with plotting / FITS output / whatever we find useful.

(Note that the equality fails whenever a `NaN` is present in the result, because `numpy.array_equal()` does not allow for NaN tolerance before 1.19 (current build has 1.17.4). `is_close()` will however work.)

### CODE REVIEW

**Comment by mlincett** [comment]
@ric-evans I wanted to put `save_scan_result.py` in a dedicated `scripts` directory, however it seems that only modules can use relative import paths. In other words, `from ..scan_result import ScanResult` would not work. For this reason I had to put `save_scan_result.py` and `scan_result.py` in the same directory (under `skymap_scanner` together with the other common files).

I am not sure how to deal with this inconvenience, maybe we should create a `setup.py` so the `skymap_scanner` package can be installed in development mode?

File: skymap_scanner/save_scan_result.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    close = result.is_close(result_check)
+    equal = result == result_check
+
+    logger.info(
+        f"The loaded file is close? ({close}) and equal? ({equal}) to the source data."
+    )
+
+
+if __name__ == "__main__":
+    main()
```
Comment (ric-evans) [inline skymap_scanner/save_scan_result.py:55]: Instead of creating a `.npz` file by loading from the `state_cache` after each run, how about we use `ScanResult` as a final step in the main pipeline? Then we compare the two `.npz` files in a dedicated test script (call it `tests/compare_scans_npz.py`).

**Review by ric-evans — CHANGES_REQUESTED** [review]
I'd prefer to not rely on `state_cache` after the pipeline has finished, where possible

**Comment by ric-evans** [comment]
> Ideally it should serve as the basic data structure for plotting / processing / transmission of the scan result.

**Comment by ric-evans** [comment]
> Ideally it should serve as the basic data structure for plotting / processing / transmission of the scan result.

I do agree with this 100%, minimizing the number of necessary outputted data files is great for loose coupling

**Comment by ric-evans** [comment]
@mlincett what're your thoughts on https://github.com/icecube/skymap_scanner/issues/9?

**Comment by mlincett** [comment]
> @mlincett what're your thoughts on #9?

The current plotting functionality and the FITS output should be integrated in `ScanResult`, but this will come later. I suggest we leave that open for now.

**Comment by mlincett** [comment]
> I'd prefer to not rely on `state_cache` after the pipeline has finished, where possible

Indeed. As we discussed, `ScanResult` should be instantiated in the pipeline and saved to a `npz` file. The provided script is meant to read the caches of previous scans (now stored in the followup machine). A new script is now included to compare two `npz` files.

The return values of this script can be changed if needed.

**Review by ric-evans — APPROVED** [review]

**Review by ric-evans — APPROVED** [review]

**Comment by ric-evans** [comment]
> When filling reco losses with zero values the test are successful. However, I would strongly advise we implement the calculation/storage of reco losses (#45) before moving on with testing the condor scan.

We'd still be able to get what we need if we persist the cache directory, correct?

**Comment by ric-evans** [comment]
Let's wait to merge this branch, so we can implement #45 here. Sound good, @mlincett?

**Comment by ric-evans** [comment]
closes #45

**Comment by mlincett** [comment]
> Let's wait to merge this branch, so we can implement #45 here. Sound good, @mlincett?

Alright. #45 is taken care of and now passing the CI test. I'd say all looks good.
